### ISSUE TITLE
TypeScript fails to narrow out `undefined` in conditional checking for that

### ISSUE THREAD

**Issue by wbt** [microsoft/TypeScript#50548]
# Bug Report
### 🔎 Search Terms
undefined narrowed 2322

### 🕗 Version & Regression Information
- This is the behavior in every version I tried, and I reviewed the [FAQ](https://github.com/Microsoft/TypeScript/wiki/FAQ) for entries about 'undefined' (including via Find on Page)

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/C4TwDgpgBAysCGAnO9gGcoF4oG8CwAUFMVALbxpoBcUAdgK6kBGEiA3ISVGvYgGbwAxhAAqEUpESpeEGg2asOREmAA28WhHRzGLdoQC+hUJCgAFdZuAp0WXJxLlKOhfuXFSAe0+1qdXYoOxJ6ITACWCKpmrGGeACYuekpcIeGRAJK0gqphtKixtImBBEYEucCsAsLmllpIIABC8WEQaACy8GD27lAA5BYaWr1QNANWNmjJJL0oiMM0sxNKpYI+aMBQAOZaNgBiiJ6kMFp2fPRZwAUAPEFQIgByUBAAHhW0cRgA1hAgnnw1gwQiEazVaHTAABpCAA+AAUt3WqGOG1GSEu8FUVzGdWBTTiLXanQA2g8ALrQqE9HLrAAyEFom2AAAtdi1VHF7vBSNAaN9fv8TBA-twENYtJSAJQjfyubpcVa+DZgTxvdGqABKrXoqg22ERYuAROpwDpDOZrIg7M53NJU2IAHp7QBBOL4hlQRAQOLnOIaDYAAwAZIGoMrVWEMZqeDqoABCTDYH0QPi5L3+qDATwZpnQBX4y4+DGUriO5nQVgHRBkVpoeDbKCCJkabYYFjAADuEHpUAAbgAWAB0AHYB32oBo4r3BwAOAcABgAtG34LcwnxYYLhWH6Wqo9qNvHsL0kynNHFelL8D0uNvaLutTq7VwoI6ABKeHuscdMD-QTPcLtUUQNUsVqIEQXxMFiTJaEiT5YVsXAvECXBEl7lJUlbhLe0AFFEErDM0FhAAmABmYjiKlFhVE8dsaDucBoH6MD6mQqCwDQ0k4J+BCWNxUFCQ4slSSgAAfKAT1Tc8oDCDBaBVcdKDCTY8iYVQ-yzQU+nkPRegHLCSEdAB1HNPSgOJPFabNUCeV4pCgf1xMks901WbkoD4A5SHHPgKirJse1yTY6CQA5O0nMsM0YqBhUi-0FXWf0AH4DOIT1gF4WhQxVHcIw1B9gCfAwnlUNBoCvZ8SHSzKoDnIrDEIIA)

### 💻 Code

```ts
type StarStats = {
    mass: number;
    surfaceTemperature: number;
    planets: number;
}
type PlanetStats = {
    mass: number;
    moons: number;
    orbitalPeriod: number;
    orbitalInclination: number;
}
interface PlanetaryBodiesMap {
    'Planet' : PlanetStats;
    'Star' : StarStats;
}
const getStatFromSet = function<
    TN extends keyof PlanetaryBodiesMap,
>(
    statSet : Partial<PlanetaryBodiesMap[TN]>,
    listLengthFieldName : keyof typeof statSet,
) : number {
    const potentialResult = statSet[listLengthFieldName];
    //Adding redundant `&& potentialResult !== undefined` to the conditional,
    //the error message changes between v4.7.4 and v4.8.0-beta
    if(typeof potentialResult !== 'undefined') {
        potentialResult;
        //Hover above to see: Partial<PlanetaryBodiesMap[TN]>[keyof PlanetaryBodiesMap[TN]]
        //Error ts(2322) below:  Type 'PlanetaryBodiesMap[TN][keyof PlanetaryBodiesMap[TN]] | undefined' is not assignable to type 'number'.
        //Where does that extra `| undefined` come from after having narrowed the type of the `const`?
        return potentialResult;
    } else {
        return 0;
    }
}
```

### 🙁 Actual behavior
TypeScript complains that a constant value can be `undefined` within a conditional block where that is already checked for and excluded.

### 🙂 Expected behavior

1. Inside a conditional which can only be entered if a constant type is not `undefined`, TypeScript excludes ` | undefined` from the possible types of the variable. 
2. When showing the type of a variable as X in an error message about how that can't be assigned to Y, the type X should be the same as the variable type shown when hovering over the type.  In this example, they differ by `| undefined`, which matters.  In the unsimplified version where the type is more complex, the type of the variable shown on hover is very linguistically different from the type shown in the error, making this issue harder to debug.

**Comment by jcalz** [comment]
This doesn't seem to have much to do with `undefined` per se; the same underlying problem happens with

```ts
const getStatFromSet = function <K extends keyof PlanetaryBodiesMap>(
    statSet: PlanetaryBodiesMap[K],
    statName: keyof typeof statSet,
): number {
    return statSet[statName]; // error
}
```

This feels like #32365 to me.

**Comment by wbt** [comment]
#32365 looks related and may have something to do with an underlying reason; it's possible that fixing that issue could also fix this one.  However, I think it's quite possible this issue could also be fixed without fixing that one by focusing on the elimination of 'undefined' from the type of any constant that is inside a conditional eliminating undefined from its type.

**Comment by jcalz** [comment]
Not seeing how, given that the following still fails:

```ts
const getStatFromSet = function <K extends keyof PlanetaryBodiesMap>(
    statSet: Partial<PlanetaryBodiesMap[K]>,
    statName: keyof typeof statSet,
): number | undefined {
    return statSet[statName]; // error!
}
```

If `statSet[statName]` cannot be seen as assignable to `number | undefined`, then eliminating `undefined` wouldn't help it being seen as assignable to `number`.  I really am not processing how `undefined` is the issue here.  Can you articulate how that would work?  Or maybe you have some more motivating code example handy?

**Comment by wbt** [comment]
The core unexpected behavior I'm trying to call attention to in this bug report is:   
```ts
  const x = //any process at all by which x gets a value, even if that's a buggy process
  if(typeof x !== 'undefined') {
        //**Regardless of how x was defined,** 
        //whether using the assignment example from the OP here or any other,
        //any attempt to use it here should NOT result in an error that "x could be `undefined`". 
  } 
```

If fixing #32365 would allow `statSet[statName]` to be assignable to type `number | undefined` in the example context given, that seems like it would be a step forward, but it's not clear to me that this would necessarily eliminate the core unexpected behavior and it's not clear that fixing the core unexpected behavior would necessarily require fixing #32365.  (It might be that fixing my problem requires fixing both, but that doesn't mean this one should be closed as a duplicate of #32365.)

**Comment by andrewbranch** [comment]
Yeah, I think there is something weird with `typeof` narrowing here. Slightly simplified to extract from #32365:

[Playground](https://www.typescriptlang.org/play?ts=4.9.0-dev.20220907#code/JYOwLgpgTgZghgYwgAgGIHt3IN4FgBQyRycAXDsgEbkDOYUoA5sgL4Fv4EwCuICYwdCGQwAPAGlkEAB6QQAExrIA1hACe6GGkwA+ABQxyABThQBcADaiM6ANriAujoA0K8qo1ab9hwEocBMTICEJ0yNLIALwitsoOANyBxMBaemBqAA4QmuHIAISR0QBEvPIQMKAQ8kX+eIRBxBYQYMgA+uTYLFHh8cgA9H3IAMoAFujcFvJSUFDoUFQQCHDcNCg5Geg0NMCUTcgABiATFvuulNwt0LNQSpSLy6vIOful5ZXy+0lEHF-BoS1qbowWIJX4pZB6QEFZCvCogKq1X5BJotdo4LrRNSJerfAi-aS9AbIAAqI2ASnJyAA7hBgFApmBML8sewCEA)

```ts
interface Foo {
    a: { b: string }
}

function f<K extends keyof Foo>(f: Partial<Foo[K]>, k: keyof Foo[K]) {
    const x = f[k];
    if (typeof x !== "undefined") {
        let _: {} = x; // Should error because of possible `null`, but errors because of `undefined`
    }

    const y = f[k];
    if (y != undefined) {
        let _: {} = y;
    }

    x; // This is weird too
    y;
}
```
